pkg/state/context.go

```diff
@@ -70,6 +71,28 @@ type State struct {
 	PauseImage                string
 }
 
+// ContainerRuntimeConfig return API object that is product of KubeOne manifest and live cluster probing
+func (s *State) ContainerRuntimeConfig() kubeoneapi.ContainerRuntimeConfig {
+	crCfg := *s.Cluster.ContainerRuntime.DeepCopy()
+	condition, _ := semver.NewConstraint(">= 1.21")
+
+	if condition.Check(s.LiveCluster.ExpectedVersion) {
+		// forced containerd for clusters version >= 1.21
+		crCfg.Docker = nil
+		crCfg.Containerd = &kubeoneapi.ContainerRuntimeContainerd{}
+		return crCfg
+	}
+
+	switch {
+	case crCfg.Docker != nil:
+	case crCfg.Containerd != nil:
+	default:
+		crCfg.Docker = &kubeoneapi.ContainerRuntimeDocker{}
+	}
+
+	return crCfg
+}
```

Comment:
I don't really like the idea that we put this information in the state. This means that we have out-of-sync information between the state and the Cluster object. If someone would, by mistake, take `s.Cluster.ContainerRuntime`, they would get incorrect information and it would be hard to catch such a bug. It would make more sense to make this something like a defaulting function (it can't be a defaulting function, but we can re-use the mechanism), i.e. make it mutate the cluster object.

Additionally, the logic is not really correct because we don't check if the cluster is provisioned or not. I believe that this check:
```go
if condition.Check(s.LiveCluster.ExpectedVersion)
```
should become:
```go
if !s.LiveCluster.IsProvisioned() && condition.Check(s.LiveCluster.ExpectedVersion)
```

Reply:
So you got me thinking again, and I think maybe we should ERROR OUT in case when the docker is specified (even of default), and plainly refuse until user will be conscious choice to change the config to indicate containerd as a container-runtime. Such approach will remove any double readings and accidents.

Reply:
I agree, but we should start doing that only from 1.22+ because 1.21 still has Docker support. To recap, the plan should be something along:
1) Default to containerd for **new** 1.21+ clusters
2) Default to containerd for **existing** 1.22+ clusters (this should be done once we have migration in-place, not now)
3) Reject Docker for **all** clusters for Kubernetes 1.22+

Reply:
OK, done, PTAL again

Reply:
This is still not fixed, see https://github.com/kubermatic/kubeone/pull/1205#discussion_r561732844.